app/code/Magento/InventoryShipping/Observer/VirtualSourceDeductionProcessor.php

```diff
@@ -0,0 +1,271 @@
+        Json $jsonSerializer,
+        InventoryRequestInterfaceFactory $inventoryRequestFactory,
+        ItemRequestInterfaceFactory $itemRequestFactory,
+        SourceSelectionServiceInterface $sourceSelectionService,
+        GetDefaultSourceSelectionAlgorithmCodeInterface $getDefaultSourceSelectionAlgorithmCode
+    ) {
+        $this->reservationBuilder = $reservationBuilder;
+        $this->appendReservations = $appendReservations;
+        $this->sourceItemsSave = $sourceItemsSave;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+        $this->getSourceItemBySourceCodeAndSku = $getSourceItemBySourceCodeAndSku;
+        $this->getSkusByProductIds = $getSkusByProductIds;
+        $this->getStockItemConfiguration = $getStockItemConfiguration;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->jsonSerializer = $jsonSerializer;
+        $this->inventoryRequestFactory = $inventoryRequestFactory;
+        $this->itemRequestFactory = $itemRequestFactory;
+        $this->sourceSelectionService = $sourceSelectionService;
+        $this->getDefaultSourceSelectionAlgorithmCode = $getDefaultSourceSelectionAlgorithmCode;
+    }
+
+    /**
+     * @param EventObserver $observer
+     * @return void
+     * @throws LocalizedException
+     */
+    public function execute(EventObserver $observer)
+    {
+        /** @var \Magento\Sales\Model\Order\Invoice $invoice */
+        $invoice = $observer->getEvent()->getInvoice();
+
+        if ($invoice->getOrigData('entity_id')) {
+            return;
+        }
+
+        $selectionRequestItems = [];
+        foreach ($invoice->getItems() as $invoiceItem) {
+            if (!$this->isValidItem($invoiceItem)) {
+                continue;
+            }
+
+            $itemSku = $invoiceItem->getSku() ?: $this->getSkusByProductIds->execute(
+                [$invoiceItem->getProductId()]
+            )[$invoiceItem->getProductId()];
+            $qty = $this->castQty($invoiceItem->getOrderItem(), $invoiceItem->getQty());
+
+            $selectionRequestItems[] = $this->itemRequestFactory->create([
+                'sku' => $itemSku,
+                'qty' => $qty,
+            ]);
+        }
+
+        if (empty($selectionRequestItems)) {
+            return;
+        }
+
+        $order = $invoice->getOrder();
+        $websiteId = $order->getStore()->getWebsiteId();
+        $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
+
+        /** @var InventoryRequestInterface $inventoryRequest */
+        $inventoryRequest = $this->inventoryRequestFactory->create([
+            'stockId' => $stockId,
+            'items' => $selectionRequestItems
+        ]);
+        $sourceSelectionResult = $this->sourceSelectionService->execute(
+            $inventoryRequest,
+            $this->getDefaultSourceSelectionAlgorithmCode->execute()
+        );
+
+        $this->deductSources($sourceSelectionResult, $stockId);
```

Comment:
after these changes current observer will look like 
```
        $invoiceItems = $invoice->getItems();
        $items = $this->someFactoryToConvertInvoiceItemsIntoSomeAbstractItems->create($invoiceItems);
        $sourceSelectionResult = $this->sourceSelection->execute($items, $stockId, $algotithmCode);
        $this->sourceDeduction->execute($sourceSelectionResult, $stockId);
```

where both of the services sourceSelection and sourceDeduction
don't have dependencies on Order/Invoice

Reply:
There are a bit more lines of code there now, but this is because we need to create an instance of `\Magento\InventorySalesApi\Api\Data\SalesEventInterface` and some `\Magento\InventoryShipping\Model\SourceDeduction\Request\SourceDeductionRequestInterface` for the given `\Magento\InventoryShipping\Model\SourceDeduction\SourceDeductionServiceInterface` implementation